File: 110.balanced-binary-tree.py
#
# @lc app=leetcode id=110 lang=python3
#
# [110] Balanced Binary Tree
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        def tree_height(root):
            if root is None:
                return 0
            left_height = tree_height(root.left)
            right_height = tree_height(root.right)

            if left_height == -1 or right_height == -1:
                return -1
            if abs(left_height - right_height) > 1:
                return -1
            return max(left_height, right_height) + 1
        if tree_height(root) == -1:
            return False
        return True
# @lc code=end

File: 110.balanced-binary-tree.ts
/*
 * @lc app=leetcode id=110 lang=typescript
 *
 * [110] Balanced Binary Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function isBalanced(root: TreeNode | null): boolean {
  let rtn: number = depth(root);
  return rtn == -1 ? false : true;
}

function depth(root: TreeNode | null): number {
  if (root === null) {
    return 0;
  }

  let left: number = depth(root.left);
  if (left == -1) {
    return -1;
  }
  let right: number = depth(root.right);
  if (right == -1) {
    return -1;
  }
  if (Math.abs(left - right) > 1) {
    return -1;
  }
  return Math.max(left, right) + 1;
}
// @lc code=end
